graph/iid/IndexIID.java

```diff
@@ -49,33 +49,123 @@
+        public static class Label extends Type {
+
+            Label(byte[] bytes) {
+                super(bytes);
+            }
+
+            /**
+             * Returns the index address of given {@code TypeVertex}
+             *
+             * @param label of the {@code TypeVertex}
+             * @param scope of the {@code TypeVertex}, which could be null
+             * @return a byte array representing the index address of a {@code TypeVertex}
+             */
+            public static Label of(String label, @Nullable String scope) {
+                return new Label(join(Encoding.Index.Prefix.TYPE.prefix().bytes(), Encoding.Vertex.Type.scopedLabel(label, scope).getBytes(STRING_ENCODING)));
+            }
+
+            @Override
+            public String toString() {
+                if (readableString == null) {
+                    readableString = "[" + PrefixIID.LENGTH + ": " + Encoding.Index.Prefix.TYPE.toString() + "]" +
+                            "[" + (bytes.length - PrefixIID.LENGTH) + ": " + new String(copyOfRange(bytes, PrefixIID.LENGTH, bytes.length), STRING_ENCODING) + "]";
+                }
+                return readableString;
+            }
         }
 
-        @Override
-        public String toString() {
-            if (readableString == null) {
-                readableString = "[" + PrefixIID.LENGTH + ": " + Encoding.Index.Prefix.TYPE.toString() + "]" +
-                        "[" + (bytes.length - PrefixIID.LENGTH) + ": " + new String(copyOfRange(bytes, PrefixIID.LENGTH, bytes.length), STRING_ENCODING) + "]";
+        public static class RuleIndex extends Type {
+
+            private boolean isScan;
+
+            RuleIndex(byte[] bytes, boolean isScan) {
+                super(bytes);
+                this.isScan = isScan;
             }
-            return readableString;
+
+            public int length() { return bytes.length; }
+
+            /**
+             * @return a byte array representing the index of a given type concluded in a given rule
+             */
+            public static RuleIndex concludedIsa(VertexIID.Type typeIID, StructureIID.Rule ruleIID) {
+                return new RuleIndex(join(Encoding.Index.Prefix.TYPE.bytes(), typeIID.bytes(),
+                                          Encoding.Index.Infix.CONCLUDED_ISA.bytes(), ruleIID.bytes()),
+                                     false);
+            }
+
+            /**
+             * @return a byte array representing the the index scan prefix of a given type concluded in rules
+             */
+            public static RuleIndex concludedIsaScan(VertexIID.Type typeIID) {
+                return new RuleIndex(join(Encoding.Index.Prefix.TYPE.bytes(), typeIID.bytes(),
+                                          Encoding.Index.Infix.CONCLUDED_ISA.bytes()),
+                                     true);
+            }
```

Comment:
this is quite useful imo and i wonder if tis applicable in other places, introducing a `scan` iid option